source/source_test.go

```diff
@@ -16,15 +16,617 @@ package source
+	}
+
+	// write files to bucket
+	addObjectsToTheBucket(ctx, t, testBucket, client, 3)
+
+	// read one record and assert position type is CDC
+	obj, err := readWithTimeout(ctx, source, time.Second*10)
+	if err != nil {
+		t.Fatal(err)
+	}
+	pos, _ := position.ParseRecordPosition(obj.Position)
+	if pos.Type != position.TypeCDC {
+		t.Fatalf("expected first position after reading an empty bucket to be CDC, got: %s", obj.Position)
+	}
+}
+
+func TestSource_CDC_ReadRecordsUpdate(t *testing.T) {
+	client, cfg := prepareIntegrationTest(t)
+
+	ctx := context.Background()
+	testBucket := cfg[config.ConfigKeyGCSBucket]
+	source := &Source{}
+	defer func() {
+		_ = source.Teardown(ctx)
+	}()
+
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = source.Open(ctx, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testFiles := addObjectsToTheBucket(ctx, t, testBucket, client, 3)
+
+	// read and assert
+	for _, file := range testFiles {
+		_, err := readAndAssert(ctx, t, source, file)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}
+
+	// make sure the update action has a different lastModifiedDate
+	// because CDC iterator detects files from after maxLastModifiedDate by initial load
+	time.Sleep(time.Second)
+
+	content := uuid.NewString()
+	testFileName := "file0000" // already exists in the bucket
+
+	wc := client.Bucket(testBucket).Object(testFileName).NewWriter(ctx)
+	fmt.Fprint(wc, content)
+	wc.Close()
+
+	obj, err := readWithTimeout(ctx, source, time.Second*10)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// the update should be detected
+	if strings.Compare(string(obj.Key.Bytes()), testFileName) != 0 {
+		t.Fatalf("expected key: %s, got: %s", testFileName, string(obj.Key.Bytes()))
+	}
+}
+
+func TestSource_CDC_ReadRecordsInsert(t *testing.T) {
+	client, cfg := prepareIntegrationTest(t)
+
+	ctx := context.Background()
+	testBucket := cfg[config.ConfigKeyGCSBucket]
+	source := &Source{}
+	defer func() {
+		_ = source.Teardown(ctx)
+	}()
+
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = source.Open(ctx, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testFiles := addObjectsToTheBucket(ctx, t, testBucket, client, 3)
+
+	// read and assert
+	for _, file := range testFiles {
+		_, err := readAndAssert(ctx, t, source, file)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}
+
+	// make sure the update action has a different lastModifiedDate
+	// because CDC iterator detects files from after maxLastModifiedDate by initial load
+	time.Sleep(time.Second)
+
+	content := uuid.NewString()
+	testFileName := "test-file"
+	// insert a file to the bucket
+	wc := client.Bucket(testBucket).Object(testFileName).NewWriter(ctx)
+	fmt.Fprint(wc, content)
+	wc.Close()
+
+	obj, err := readWithTimeout(ctx, source, time.Second*15)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// the insert should have been detected
+	if strings.Compare(string(obj.Key.Bytes()), testFileName) != 0 {
```

Comment:
How about the payload?

Reply:
Seems the key is enough because we required whether it captured the exact action/event. Yes now i have also validating the content to make sure the data is updated correctly and same is captured.

Reply:
Ok Now we are validating the payload too.